### PR TITLE
close #9 - add ability to cancel unflip

### LINKED ISSUE #9: Add an option to cancel unflipping
Needs to be done if unflipping time option would allow bigger values.

### PR DESCRIPTION
**Pull request by McDiod**
* needs CBA >= 3.12.0
* makes unflip progressbar closeable
* sends stop event to server if unflip progressbar is closed
* checks continuously for _requiredUnits both on server and client

### CODE REVIEW

File: addons/unflipping/XEH_preInitClient.sqf [hunk 1]
```diff
@@ -8,23 +9,52 @@
...
         [
             localize "STR_vet_unflipping_doing",
-            _this,
-            {true},
-            {},
+            // time
+            _this#2,
+            // condition
+            {
+                params ["_args","","_elapsedTime"];
```
Comment (veteran29) [inline addons/unflipping/XEH_preInitClient.sqf]:
```suggestion
                params ["_args", "", "_elapsedTime"];
```

File: addons/unflipping/XEH_preInitClient.sqf [hunk 2]
```diff
@@ -8,23 +9,52 @@
...
+                // don't check before 1s elapsed to wait for publicVariable synchronization
+                _elapsedTime < 1 ||
+                {count (_vehicle getVariable ["vet_unflippingUnits", []]) >= _requiredUnits}
+            },
+            // onSuccess
             {},
-            [],
+            // onFailure
+            {
+                params ["_args","","","","_failureCode"];
```
Comment (veteran29) [inline addons/unflipping/XEH_preInitClient.sqf]:
```suggestion
                params ["_args", "", "", "", "_failureCode"];
```

File: addons/unflipping/XEH_preInitClient.sqf [hunk 3]
```diff
@@ -8,23 +9,52 @@
...
+
+                // user hit ESC
+                if (_failureCode == 1) then {
+                    ["vet_unflipping_unflip_stop", [_vehicle, PLAYER]] call CBA_fnc_serverEvent;
+
+                // user did not hit ESC --> other reason for failure
+                } else {
+                    // if user is in unflippingUnits --> enter wait mode again
+                    // if not --> server has completed unflipping and reset the array
+                    if (PLAYER in (_vehicle getVariable ["vet_unflippingUnits",[]])) then {
```
Comment (veteran29) [inline addons/unflipping/XEH_preInitClient.sqf]:
```suggestion
                    if (PLAYER in (_vehicle getVariable ["vet_unflippingUnits", []])) then {
```

File: addons/unflipping/XEH_preInitServer.sqf [hunk 4]
```diff
@@ -14,13 +14,30 @@
...
+
         // Schedule unflip
-        [{
-            _this call vet_unflipping_fnc_unflipVehicle;
-            _this setVariable ["vet_unflippingUnits", [], true];
-        }, _vehicle, vet_unflipping_time] call CBA_fnc_waitAndExecute;
+        [
+            // condition
+            {
+                params ["_vehicle","_requiredUnits"];
```
Comment (veteran29) [inline addons/unflipping/XEH_preInitServer.sqf]:
```suggestion
                params ["_vehicle", "_requiredUnits"];
```

File: addons/unflipping/XEH_preInitServer.sqf [hunk 5]
```diff
@@ -14,13 +14,30 @@
...
+                count (_vehicle getVariable ["vet_unflippingUnits", []]) < _requiredUnits
+            },
+            // statement (failure)
+            {},
+            // args
+            [_vehicle,_requiredUnits],
```
Comment (veteran29) [inline addons/unflipping/XEH_preInitServer.sqf]:
```suggestion
            [_vehicle, _requiredUnits],
```

File: addons/unflipping/functions/fn_unflipAction.sqf [hunk 6]
```diff
@@ -50,22 +50,29 @@ if !(PLAYER in UNFLIPPING_UNITS) exitWith {
...
             [
                 ["\a3\3den\data\attributes\loiterdirection\cw_ca.paa"],
                 [localize "STR_vet_unflipping_need_more"]
             ] call CBA_fnc_notify;
         },
+        // onFailure
         {
-            params ["_vehicle"];
-            ["vet_unflipping_unflip_stop", [_vehicle, PLAYER]] call CBA_fnc_serverEvent;
+            params ["_args","","","","_failureCode"];
```
Comment (veteran29) [inline addons/unflipping/functions/fn_unflipAction.sqf]:
```suggestion
            params ["_args", "", "", "", "_failureCode"];
```

**Review by veteran29 — CHANGES_REQUESTED** [review]
Thank you very much for this! Looks robust. 

I've only some code-style related suggestions.

**Comment by McDiod** [comment]
Oh yea, I tried to match your style, but apparently forgot about the array spaces along the way.

**Pull request merged**